odml/tools/rdf_converter.py

```diff
@@ -34,147 +64,187 @@ class RDFWriter(object):
         """
-        fmt = e.format()
+        save_odml_list adds all odml elements in a list to the current
+        parent node and handles all child items via save_element.
 
-        if not node:
-            curr_node = URIRef(odmlns + unicode(e.id))
-        else:
-            curr_node = node
+        :param odml_list: list of odml entities.
+        :param parent_node: current parent node in the RDF graph.
+        :param rdf_predicate: RDF predicate used to add all odml entities
+                              to the parent node.
+        """
+        for curr_item in odml_list:
+            node = URIRef(ODML_NS + unicode(curr_item.id))
+            self.graph.add((parent_node, rdf_predicate, node))
+            self.save_element(curr_item, node)
 
-        if fmt.name == "section":
-            s = self._get_section_subclass(e)
-            u = s if s else fmt.rdf_type
-            self.g.add((curr_node, RDF.type, URIRef(u)))
-        else:
-            self.g.add((curr_node, RDF.type, URIRef(fmt.rdf_type)))
+    def save_repository_node(self, parent_node, rdf_predicate, leaf_value):
+        """
+        save_repository_node adds a node with a given repository url to
+        the current graphs terminology node. If the current graph does
+        not yet contain a terminology node, it creates one and attaches
+        the current node to it.
+
+        :param parent_node: current parent node in the RDF graph.
+        :param rdf_predicate: RDF predicate used to add the terminology
+                              to the parent node.
+        :param leaf_value: Value that will be added to the RDF graph.
+        """
+        terminology_node = self.graph.value(predicate=RDF.type, object=URIRef(leaf_value))
+        if not terminology_node:
+            # adding terminology url value to the graph and linking it
+            # to the current RDF node.
+            terminology_node = URIRef(ODML_NS + unicode(uuid.uuid4()))
+            self.graph.add((terminology_node, RDF.type, URIRef(leaf_value)))
+            self.graph.add((self.hub_root, ODML_NS.hasTerminology, terminology_node))
+
+        self.graph.add((parent_node, rdf_predicate, terminology_node))
 
-        # adding doc to the hub
-        if isinstance(fmt, Document.__class__):
-            self.g.add((self.hub_root, odmlns.hasDocument, curr_node))
+    def save_element(self, odml_elem, node=None):
+        """
+        Save the current odml element to the RDF graph and handle all child
+        elements of the current odml element recursively.
 
-            # If available add the documents filename to the document node
+        :param odml_elem: An odml element that should be added to the RDF graph.
+        :param node: An RDF node that is used to append the current odml element
+                     to the RDF graph. If None, a new node will be created and
+                     added to the 'Hub' node of the RDF graph.
+        """
```

Comment:
this function is a monster ;) Might be cool to split is up to functions that handle the different subtype separately. Something for the future, maybe.

Reply:
I agree, the method is still too big and the suggestion to split it up into subtype handling intrigues me. Lets see what I can do about it...